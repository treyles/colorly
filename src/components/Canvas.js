import React from 'react';
import { storage } from '../utils/base';

export default class Canvas extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      mousePosition: {},
      imageLoaded: false,
      previewColor: false,
      currentColor: null,
      winWidth: null,
      winHeight: null
    };

    this.handleColorAndPosition = this.handleColorAndPosition.bind(this);
    this.handleCanvasLeave = this.handleCanvasLeave.bind(this);
    this.handleColorPick = this.handleColorPick.bind(this);
    this.updateWindowDimensions = this.updateWindowDimensions.bind(this);
  }

  componentDidMount() {
    this.ctx = this.canvas.getContext('2d');
    this.updateWindowDimensions();
    window.addEventListener('resize', this.updateWindowDimensions);
    this.getImageFile();
  }

  updateWindowDimensions() {
    this.setState({
      winWidth: window.innerWidth,
      winHeight: window.innerHeight
    });
  }

  // TODO: name handleColorPreview ???
  handleColorAndPosition(e) {
    const { imageLoaded } = this.state;

    const rect = this.canvas.getBoundingClientRect();
    const actualX = Math.floor(e.clientX - rect.left);
    const actualY = Math.floor(e.clientY - rect.top);

    const pixelData = this.ctx.getImageData(actualX, actualY, 1, 1);
    const { data } = pixelData;
    const pixelColor = `rgba(${data[0]}, ${data[1]}, ${data[2]}, ${data[3]})`;

    if (imageLoaded) {
      this.setState({
        currentColor: pixelColor,
        mousePosition: { x: actualX, y: actualY },
        previewColor: true
      });
    }
  }

  handleCanvasLeave() {
    this.setState({
      previewColor: false
    });
  }

  handleColorPick() {
    const { currentColor } = this.state;
    this.props.makePalette(currentColor);
  }

  drawCanvasImage(image) {
    const heightPadding = 275;
    const widthPadding = 60;
    const containerWidth = this.state.winWidth - widthPadding;
    const containerHeight = this.state.winHeight - heightPadding;
    let imageWidth = image.width;
    let imageHeight = image.height;

    // if any part of the image (original dimensions) overflow
    // the container, shrink dimensions and maintain aspect ratio
    if (imageWidth > containerWidth || imageHeight > containerHeight) {
      // algorithm modified from 1owk3y: https://bit.ly/2JiYzHz
      const wFits = containerWidth / imageWidth;
      const hFits = containerHeight / imageHeight;
      const minFits = wFits > hFits ? hFits : wFits;

      imageWidth *= minFits;
      imageHeight *= minFits;
    }

    this.canvas.width = imageWidth;
    this.canvas.height = imageHeight;

    // TODO: check if this works
    this.ctx.webkitImageSmoothingEnabled = false;
    this.ctx.drawImage(image, 0, 0, imageWidth, imageHeight);
  }

  getImageFile() {
    const imageFile = storage
      .ref()
      .child(`images/${this.props.imageName}`);

    imageFile.getDownloadURL().then(url => {
      const image = new Image();
      image.crossOrigin = 'anonymous';

      image.onload = () => {
        this.drawCanvasImage(image);
        this.setState({ imageLoaded: true });
      };

      image.src = url;
    });
  }

  render() {
    const { mousePosition, previewColor, currentColor } = this.state;

    const containerStyle = {
      top: `${mousePosition.y}px`,
      left: `${mousePosition.x + 10}px`
    };

    const colorStyle = {
      background: currentColor
    };

    return (
      <div className="canvas-wrapper">
        <canvas
          ref={ref => {
            this.canvas = ref;
          }}
          // width={800}
          // height={550}
          onMouseMove={this.handleColorAndPosition}
          onMouseLeave={this.handleCanvasLeave}
          onClick={this.handleColorPick}
        />
        {previewColor && (
          <div className="preview-container" style={containerStyle}>
            <div className="preview-color" style={colorStyle} />
          </div>
        )}
      </div>
    );
  }
}
