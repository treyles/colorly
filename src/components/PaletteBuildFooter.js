import React from 'react';
import PropTypes from 'prop-types';
import { withRouter } from 'react-router-dom';
import { connect } from 'react-redux';
import Icon from '../utils/Icon';
import {
  setCheckedColor,
  setPaletteTitle,
  clearPalette,
  addCardToLibrary
} from '../actions';

class PaletteBuildFooter extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      submitAlert: false
    };

    this.handleTitleIconClick = this.handleTitleIconClick.bind(this);
    this.handleKeyDown = this.handleKeyDown.bind(this);
    this.setTitleInputRef = this.setTitleInputRef.bind(this);
    this.handleSaveClick = this.handleSaveClick.bind(this);
  }

  componentWillUnmount() {
    clearTimeout(this.submitAlertTimer);
  }

  setTitleInputRef(ref) {
    this.titleInput = ref;
  }

  handleTitleIconClick() {
    this.titleInput.focus();
  }

  handleKeyDown(e) {
    if (e.keyCode === 13) {
      this.titleInput.blur();
    }
  }

  handleSaveClick() {
    const { palette, title, history } = this.props;

    if (!Object.keys(palette).length || !title.length) {
      this.activateSubmitAlert();
      return;
    }

    this.props.addCardToLibrary(history);
  }

  activateSubmitAlert() {
    const { palette } = this.props;
    let { submitAlert } = this.state;

    if (!Object.keys(palette).length) {
      submitAlert = 'Palette Is Empty!';
    } else {
      submitAlert = 'Title Required!';
    }

    this.setState({ submitAlert });
    this.resetSubmitAlert();
  }

  resetSubmitAlert() {
    this.submitAlertTimer = setTimeout(() => {
      this.setState({ submitAlert: false });
    }, 1500);
  }

  // TODO: save for last (if have time), loop palette color markup
  render() {
    const { submitAlert } = this.state;
    const {
      palette,
      checked,
      title,
      library,
      clearPalette,
      setCheckedColor,
      setPaletteTitle
    } = this.props;

    return (
      <div className="footer">
        <div className="palette-name">
          <span
            onClick={this.handleTitleIconClick}
            onKeyDown={this.handleTitleIconClick}
            role="button"
            tabIndex="0"
          >
            <Icon icon="pencil" />
          </span>
          <input
            ref={this.setTitleInputRef}
            onChange={e => setPaletteTitle(e)}
            value={title}
            className="palette-input"
            type="text"
            placeholder={`Color Palette #${library.length + 1}`}
            maxLength="24"
            onKeyDown={this.handleKeyDown}
          />
        </div>
        <div className="palette-colors">
          <input
            type="radio"
            id="color1"
            checked={checked === 'color1'}
            onChange={e => setCheckedColor(e)}
          />
          <label
            className="radio"
            htmlFor="color1"
            style={{ background: palette.color1 }}
          />
          <input
            type="radio"
            id="color2"
            checked={checked === 'color2'}
            onChange={e => setCheckedColor(e)}
          />
          <label
            className="radio"
            htmlFor="color2"
            style={{ background: palette.color2 }}
          />
          <input
            type="radio"
            id="color3"
            checked={checked === 'color3'}
            onChange={e => setCheckedColor(e)}
          />
          <label
            className="radio"
            htmlFor="color3"
            style={{ background: palette.color3 }}
          />
          <input
            type="radio"
            id="color4"
            checked={checked === 'color4'}
            onChange={e => setCheckedColor(e)}
          />
          <label
            className="radio"
            htmlFor="color4"
            style={{ background: palette.color4 }}
          />
          <input
            type="radio"
            id="color5"
            checked={checked === 'color5'}
            onChange={e => setCheckedColor(e)}
          />
          <label
            className="radio"
            htmlFor="color5"
            style={{ background: palette.color5 }}
          />
        </div>
        <div className="palette-buttons">
          <button className="clear" onClick={() => clearPalette()}>
            RESET
          </button>
          <button className="save" onClick={this.handleSaveClick}>
            SAVE PALETTE
            <div className={`submit-alert ${submitAlert ? 'active' : ''}`}>
              {submitAlert}
            </div>
          </button>
        </div>
      </div>
    );
  }
}

PaletteBuildFooter.propTypes = {
  history: PropTypes.object.isRequired, // eslint-disable-line
  // action creators
  setCheckedColor: PropTypes.func.isRequired,
  setPaletteTitle: PropTypes.func.isRequired,
  clearPalette: PropTypes.func.isRequired,
  addCardToLibrary: PropTypes.func.isRequired,
  // from store
  library: PropTypes.arrayOf(PropTypes.object).isRequired,
  checked: PropTypes.string.isRequired,
  palette: PropTypes.objectOf(PropTypes.string).isRequired,
  title: PropTypes.string.isRequired
};

const mapStateToProps = state => ({
  library: state.data.library,
  checked: state.build.checked,
  palette: state.build.card.palette,
  title: state.build.card.title
});

export default withRouter(
  connect(mapStateToProps, {
    setCheckedColor,
    setPaletteTitle,
    clearPalette,
    addCardToLibrary
  })(PaletteBuildFooter)
);
