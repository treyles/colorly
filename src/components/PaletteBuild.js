import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import PaletteBuildFooter from './PaletteBuildFooter';
import BackButton from './BackButton';
import Canvas from './Canvas';
import DropZoneContent from './DropZoneContent';
import { resetBuild, setImageSource } from '../actions';

export class PaletteBuild extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      dropZoneHover: false,
      imageAlert: false
    };

    this.handleImageDrop = this.handleImageDrop.bind(this);
    this.handleDragOver = this.handleDragOver.bind(this);
    this.handleDragLeave = this.handleDragLeave.bind(this);
  }

  componentWillUnmount() {
    clearTimeout(this.imageAlertTimer);
    this.props.resetBuild();
  }

  handleDragOver(e) {
    e.preventDefault();
    this.setState({ dropZoneHover: true });
  }

  handleDragLeave(e) {
    e.preventDefault();
    this.setState({ dropZoneHover: false });
  }

  handleImageDrop(e) {
    e.preventDefault();
    // change scope based on image 'dropped' or 'selected'
    const imageFile = e.dataTransfer
      ? e.dataTransfer.files[0]
      : e.target.files[0];
    // 3 megabyte limit
    const sizeLimit = 1024 * 3;
    const fileSize = imageFile.size / 1024;
    const validTypes = [
      'image/jpeg',
      'image/gif',
      'image/png',
      'image/svg+xml'
    ];

    // check for valid image
    if (!validTypes.includes(imageFile.type)) {
      this.setState({
        imageAlert: 'Not a Valid File Type!',
        dropZoneHover: false
      });

      this.resetImageLoadAlert();
      return;
    }

    if (fileSize > sizeLimit) {
      this.setState({
        imageAlert: 'Image Must be Smaller Than 3 MB!',
        dropZoneHover: false
      });

      this.resetImageLoadAlert();
      return;
    }

    this.props.setImageSource(imageFile);
  }

  resetImageLoadAlert() {
    this.imageAlertTimer = setTimeout(() => {
      this.setState({ imageAlert: false });
    }, 1500);
  }

  render() {
    const { dropZoneHover, imageAlert } = this.state;
    const { imageSource, showPreloader } = this.props;

    const preloader = (
      <div className="preloader-wrapper">
        <span className="preloader">Loading</span>
      </div>
    );

    return (
      <div className="palette-build">
        <Link to="/">
          <BackButton />
        </Link>
        {!imageSource ? (
          <div
            className={`drop-zone ${dropZoneHover ? 'dragover' : ''}`}
            onDragOver={this.handleDragOver}
            onDragLeave={this.handleDragLeave}
            onDrop={this.handleImageDrop}
          >
            <DropZoneContent
              imageAlert={imageAlert}
              handleImageDrop={this.handleImageDrop}
            />
          </div>
        ) : (
          <div className="canvas-container">
            <Canvas imageSource={imageSource} />
            <PaletteBuildFooter />
          </div>
        )}
        {showPreloader && preloader}
      </div>
    );
  }
}

PaletteBuild.propTypes = {
  resetBuild: PropTypes.func.isRequired,
  setImageSource: PropTypes.func.isRequired,
  imageSource: PropTypes.oneOfType([PropTypes.bool, PropTypes.object])
    .isRequired,
  showPreloader: PropTypes.bool.isRequired
};

const mapStateToProps = state => ({
  imageSource: state.build.imageSource,
  showPreloader: state.build.showPreloader
});

export default connect(mapStateToProps, { resetBuild, setImageSource })(
  PaletteBuild
);
