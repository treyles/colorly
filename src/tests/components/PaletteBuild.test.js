import React from 'react';
import { shallow } from 'enzyme';
import PaletteBuild from '../../components/PaletteBuild';
import Canvas from '../../components/Canvas';
import PaletteBuildFooter from '../../components/PaletteBuildFooter';

describe('PaletteBuild', () => {
  let component;
  jest.useFakeTimers();

  const props = {
    library: [],
    currentUser: true,
    addCardToLibrary: jest.fn(),
    history: {}
  };

  const event = {
    preventDefault: jest.fn(),
    target: {
      files: [
        {
          type: 'image/jpeg',
          size: 125713
        }
      ]
    }
  };

  beforeEach(() => {
    component = shallow(<PaletteBuild {...props} />);
  });

  it('renders correctly', () => {
    expect(component).toMatchSnapshot();
  });

  it('alerts to invalid file TYPE when image is dropped/selected', () => {
    event.target.files[0].type = 'image/notajpeg';
    component.instance().handleImageDrop(event);
    expect(component.state().imageAlert).toEqual('Not a Valid File Type!');
  });

  it('alerts to invalid file SIZE when image is dropped/selected', () => {
    event.target.files[0] = {
      type: 'image/jpeg',
      size: 8330244
    };
    component.instance().handleImageDrop(event);
    expect(component.state().imageAlert).toEqual(
      'Image Must be Smaller Than 3 MB!'
    );
  });

  it('resets imageAlert after 1.5s', () => {
    component.instance().handleImageDrop(event);
    expect(component.state().imageAlert).toEqual(
      'Image Must be Smaller Than 3 MB!'
    );
    setTimeout(() => {
      expect(component.state().imageAlert).toEqual(false);
    }, 1500);
    jest.runAllTimers();
  });

  it('adds new color to palette state when makePalette is called', () => {
    component.instance().makePalette('rgbValue');
    expect(component.state().palette.color1).toEqual('rgbValue');
  });

  it('clears palette when clearPalette is called ', () => {
    component.setState({
      palette: {
        color1: 'palette1'
      }
    });
    expect(Object.keys(component.state().palette).length).toEqual(1);
    component.instance().clearPalette();
    expect(Object.keys(component.state().palette).length).toEqual(0);
  });
  it('renders <Canvas /> if imageSource is truthy', () => {
    component.setState({
      imageSource: 'url'
    });
    expect(component.find(Canvas).exists()).toEqual(true);
  });

  it('renders <PaletteBuildFooter /> if imageSource is truthy', () => {
    component.setState({
      imageSource: 'url'
    });
    expect(component.find(PaletteBuildFooter).exists()).toEqual(true);
  });

  it('renders drop-zone if imageSource is falsey', () => {
    expect(component.find('.drop-zone').exists()).toEqual(true);
  });

  it('renders preloader if showPreloader is truthy', () => {
    component.setState({
      showPreloader: true
    });
    expect(component.find('.preloader-wrapper').exists()).toEqual(true);
  });
});
