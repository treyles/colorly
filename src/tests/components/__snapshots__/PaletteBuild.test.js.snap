// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PaletteBuild renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <PaletteBuild
    addCardToLibrary={[MockFunction]}
    currentUser={true}
    history={Object {}}
    library={Array []}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Link
          replace={false}
          to="/"
        >
          <BackButton
            onClick={null}
          />
        </Link>,
        <div
          className="drop-zone "
          onDragLeave={[Function]}
          onDragOver={[Function]}
          onDrop={[Function]}
        >
          <DropZoneContent
            handleImageDrop={[Function]}
            imageAlert={false}
          />
        </div>,
        false,
      ],
      "className": "palette-build",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": <BackButton
            onClick={null}
          />,
          "replace": false,
          "to": "/",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "onClick": null,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <DropZoneContent
            handleImageDrop={[Function]}
            imageAlert={false}
          />,
          "className": "drop-zone ",
          "onDragLeave": [Function],
          "onDragOver": [Function],
          "onDrop": [Function],
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "handleImageDrop": [Function],
            "imageAlert": false,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": "div",
      },
      false,
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Link
            replace={false}
            to="/"
          >
            <BackButton
              onClick={null}
            />
          </Link>,
          <div
            className="drop-zone "
            onDragLeave={[Function]}
            onDragOver={[Function]}
            onDrop={[Function]}
          >
            <DropZoneContent
              handleImageDrop={[Function]}
              imageAlert={false}
            />
          </div>,
          false,
        ],
        "className": "palette-build",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": <BackButton
              onClick={null}
            />,
            "replace": false,
            "to": "/",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "onClick": null,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <DropZoneContent
              handleImageDrop={[Function]}
              imageAlert={false}
            />,
            "className": "drop-zone ",
            "onDragLeave": [Function],
            "onDragOver": [Function],
            "onDrop": [Function],
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "handleImageDrop": [Function],
              "imageAlert": false,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": "div",
        },
        false,
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getSnapshotBeforeUpdate": true,
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
